ddtrace/internal/ci_visibility/encoder.py

```diff
@@ -0,0 +1,68 @@
+import threading
+from typing import Any
+from typing import Dict
+from typing import TYPE_CHECKING
+
+from .._encoding import BufferedEncoder
+from .._encoding import packb as msgpack_packb
+from ..encoding import JSONEncoderV2
+
+
+if TYPE_CHECKING:  # pragma: no cover
+    from ..span import Span
+
+
+class CIVisibilityEncoderV01(BufferedEncoder):
+    content_type = "application/msgpack"
+    ALLOWED_METADATA_KEYS = ("language", "library_version", "runtime-id", "env")
+    PAYLOAD_FORMAT_VERSION = 1
+
+    def __init__(self, *args):
+        super(CIVisibilityEncoderV01, self).__init__()
+        self._lock = threading.RLock()
+        self._init_buffer()
+
+    def __len__(self):
+        with self._lock:
+            return len(self.buffer)
+
+    def set_metadata(self, metadata):
+        self._metadata = metadata or dict()
+
+    def _init_buffer(self):
+        with self._lock:
+            self.buffer = []
+
+    def put(self, spans):
+        with self._lock:
+            self.buffer.append(spans)
+
+    def encode_traces(self, traces):
+        return self._build_payload(traces=traces)
+
+    def encode(self):
+        with self._lock:
+            payload = self._build_payload(self.buffer)
+            self._init_buffer()
+            return payload
+
+    def _build_payload(self, traces):
+        normalized_spans = [CIVisibilityEncoderV01._convert_span(span) for trace in traces for span in trace]
+        self._metadata = {k: v for k, v in self._metadata.items() if k in self.ALLOWED_METADATA_KEYS}
+        return msgpack_packb(
```

Comment:
```suggestion
        # TODO: Split the events in several payloads as needed to avoid hitting the intake's maximum payload size.
        return msgpack_packb(
```

This logic makes builidng the messagepack a bit more convoluted (since you don't really know how many elements you will have in the array ahead of time) so we don't have to fix this now.

Reply:
We can also track this internally in a Jira task, if you don't want to merge `TODO` comments.